### ISSUE TITLE
ipython console does not connect to kernel

### ISSUE THREAD

**Issue by amrit-poudel** [spyder-ide/spyder#2754]
I installed a precompiled spyder2 (2.3.7) in my macosx 10.10. 

I understand that spyder2 comes with its own ipython, python and numpy/scipy stack, but I just want to use my own python executable and numpy/scipy  along with many other python libraries that spyder does not include.

I compiled all python libraries (numpy scipy, matplotlib,  ipython, h5py  etc. ) from the source and from the preference menu on spyder, I have set the PYTHONPATH to point to my version of python libraries and in the advanced console, I have set my version of python executable.

However, ipython gets stuck and does not connect to the kernel.  Running ipython on the teminal works fine, so there is no problem with my installation of ipython, which I compiled from the source.

If I use ipython that came with spyder2, it works fine and connects to the kernel easily. However, if I switch to my ipython in spyder2, it cannot connect to the kernel.

I do not understand why it does that. Any help will be appreciated.

**Comment by ccordoba12** [comment]
Adding things to PYTHONPATH is not the right way to accomplish what you want. Please see this answer I gave in StackOverflow:

http://stackoverflow.com/a/19813292/438386

to understand what you need to do.

In a few words: you need to change your Python interpreter to point to the one you have all the libraries you want to use. The details are in the answer :-)

**Comment by amrit-poudel** [comment]
Thanks for your quick reply. 

I have obviously tried 

Open Spyder and go to

Tools > Preferences > Console > Advanced settings > Python Executable

Select the option

Use the following Python interpreter


This does not seem to help.

**Comment by amrit-poudel** [comment]
Thanks for your reply.

I tried your suggestion, but I still cannot get ipython console connect to
the kernel.

I will appreciate any other suggestions.

Thanks.

Amirt

**Comment by amrit-poudel** [comment]
Is it required to have ipythone qtconsole to get the ipython console connect with kernel in spyder?  I do not think I have ipython qtconsole installed.  I have ipythone 4.0 installed, but I do not see ipython qtconsole.

**Comment by ccordoba12** [comment]
Yes, you need to run

    pip install qtconsole

to have it in Homebrew.

**Comment by amrit-poudel** [comment]
I installed qtconsole from the source. However, I still get the same error: 

Connecting to kernel...

**Comment by amrit-poudel** [comment]
I am getting this error in spyder. I do not get this error when I run ipython directly in the terminal.

853 
    854         # Set all default hooks, defined in the IPython.hooks module.
--> 855         hooks = IPython.core.hooks
        hooks = undefined
        global IPython.core.hooks = undefined
    856         for hook_name in hooks.__all__:
    857             # default hooks have priority 100, i.e. low; user hooks should have

AttributeError: 'module' object has no attribute 'core'

**Comment by ccordoba12** [comment]
This seems an installation problem that doesn't have to do with Spyder. Please run

    pip install -U qtconsole

and then

    pip install -U jupyter

If that doesn't work, I don't know how else to help you.

**Comment by amrit-poudel** [comment]
Thanks.

First I installed jupyter and when I found out that ipython console was not connecting to the kernel in spyder, I then installed qtconsole.

I put everything in the same folder.

Does the ordering of the installation matter?


Please note that I am not using homebrew. I do regular installation using: 

cd jupyter
python setup.py build
python setup.py install --prefix=/somefolder

And the same for qtconsole-4.0.0

I am pretty sure it is spyder related, because ipython works fine on the terminal.  ipython does not connect to the kernel only when I try to run inside spyder.

**Comment by amrit-poudel** [comment]
I do not understand this error in spyder. I am unable to use ipython in spyder. On the terminal, ipython runs fine, however.

AttributeError Traceback (most recent call last)
/Users/lib/python2.7/site‑packages/spyderlib/widgets/externalshell/start_ipython_kernel.py in ()
184 ipk_temp = IPKernelApp.instance()
185 ipk_temp.config = kernel_config()
‑‑> 186 ipk_temp.initialize()
global ipk_temp.initialize = >
187 
188 # Grabbing the kernel's shell to share its namespace with our

/Users/lib/python2.7/site‑packages/ipykernel/kernelapp.pyc in initialize(self=, argv=None)

/Users/lib/python2.7/site‑packages/traitlets/config/application.pyc in catch_config_error(method=, app=, *args=(None,), **kwargs={})
73 """
74 try:
‑‑‑> 75 return method(app, *args, **kwargs)
method = 
app = 
args = (None,)
kwargs = {}
76 except (TraitError, ArgumentError) as e:
77 app.print_help()

/Users/lib/python2.7/site‑packages/ipykernel/kernelapp.pyc in initialize(self=, argv=None)
380 self.init_io()
381 self.init_signal()
‑‑> 382 self.init_kernel()
self.init_kernel = >
383 # shell init steps
384 self.init_path()

/Users//lib/python2.7/site‑packages/ipykernel/kernelapp.pyc in init_kernel(self=)
319 log=self.log,
320 profile_dir=self.profile_dir,
‑‑> 321 user_ns=self.user_ns,
global user_ns = undefined
self.user_ns = None
322 )
323 kernel.record_ports(self.ports)

/Users/lib/python2.7/site‑packages/traitlets/config/configurable.pyc in instance(cls=, *args=(), **kwargs={'iopub_socket': , 'log': , 'parent': , 'profile_dir': , 'session': , 'shell_streams': [, ], 'stdin_socket': , 'user_ns': None})
378 # Create and save the instance
379 if cls._instance is None:
‑‑> 380 inst = cls(*args, **kwargs)
inst = undefined
cls = 
args = ()
kwargs = {'log': , 'parent': , 'shell_streams': [, ], 'user_ns': None, 'iopub_socket': , 'profile_dir': , 'session': , 'stdin_socket': }
381 # Now make sure that the instance will also be returned by
382 # parent classes' _instance attribute.

/Users/lib/python2.7/site‑packages/ipykernel/ipkernel.pyc in __init__(self=, **kwargs={'iopub_socket': , 'log': , 'parent': , 'profile_dir': , 'session': , 'shell_streams': [, ], 'stdin_socket': , 'user_ns': None})
45 user_module = self.user_module,
46 user_ns = self.user_ns,
‑‑‑> 47 kernel = self,
global kernel = undefined
self = 
48 )
49 self.shell.displayhook.session = self.session

/Users/lib/python2.7/site‑packages/traitlets/config/configurable.pyc in instance(cls=, *args=(), **kwargs={'kernel': , 'parent': , 'profile_dir': , 'user_module': None, 'user_ns': None})
378 # Create and save the instance
379 if cls._instance is None:
‑‑> 380 inst = cls(*args, **kwargs)
inst = undefined
cls = 
args = ()
kwargs = {'kernel': , 'user_ns': None, 'profile_dir': , 'parent': , 'user_module': None}
381 # Now make sure that the instance will also be returned by
382 # parent classes' _instance attribute.

/Users/lib/python2.7/site‑packages/IPython/core/interactiveshell.pyc in __init__(self=, ipython_dir=None, profile_dir=, user_module=None, user_ns=None, custom_exceptions=((), None), **kwargs={'kernel': , 'parent': })
516 
517 self.init_syntax_highlighting()
‑‑> 518 self.init_hooks()
self.init_hooks = >
519 self.init_events()
520 self.init_pushd_popd_magic()

/Users/lib/python2.7/site‑packages/ipykernel/zmqshell.pyc in init_hooks(self=)
400 
401 def init_hooks(self):
‑‑> 402 super(ZMQInteractiveShell, self).init_hooks()
global super = undefined
global ZMQInteractiveShell = 
self.init_hooks = >
403 self.set_hook('show_in_pager', page.as_hook(payloadpage.page), 99)
404 

/Users/lib/python2.7/site‑packages/IPython/core/interactiveshell.pyc in init_hooks(self=)
853 
854 # Set all default hooks, defined in the IPython.hooks module.
‑‑> 855 hooks = IPython.core.hooks
hooks = undefined
global IPython.core.hooks = undefined
856 for hook_name in hooks.__all__:
857 # default hooks have priority 100, i.e. low; user hooks should have

AttributeError: 'module' object has no attribute 'core'

**Comment by ccordoba12** [comment]
@amrit-poudel, you're not following my instructions correctly. I told you to run

    pip install -U jupyter

and you ran

    python setup.py build
    python setup.py install --prefix=/somefolder

on the jupyter package.

Those two things are not equivalent because the last one doesn't pull the packages jupyter depends on, it just installs its contents and no more.

So please use the command I gave you and report back :-)

**Comment by amrit-poudel** [comment]
Could you please tell me exactly what other packages jupyter depends on? 

I want to avoid using pip install and use python setup.py build and install (which is the standard way of building any python packages from source. Besides I do not have pip installed).

FYI, I have following packages already installed (I did not install them in that order, but I have them all installed and are in pythonpath):

decorator
gnureadline
ipykernel
ipython_genutils
jupyter-core
jupyter-client
pexpect
pickleshare
ptyprocess
pyzmq
qtconsole
requests
simplegeneric
testpaths
traitlets
zmq


What else do I need to install to make ipython console work on spyder? 


Is pip install  the *only* way to install jupyter?

**Comment by ccordoba12** [comment]
I'm afraid I'm going to close this issue. And please don't open new ones. 

`pip` is used by the Python community at large to install packages. It's main advantage is that it pulls all dependencies a package depends on.

It's totally fine if you don't want to use it, but that's a **non-standard** installation procedure, so you're on your own to solve all problems that could appear in your way.

Please note that this situation doesn't represent a bug in Spyder, it's just a problem with the way you're trying to install its dependencies. So that's another reason for me to close this issue.

**Comment by amrit-poudel** [comment]
Could you at least tell me what are the dependencies for jupyter? Instead of randomly installing arbitrary package via pip, I was trying to understand more systematically, what was causing for ipython to break  in spyder.

**Comment by blink1073** [comment]
@amrit-poudel, there is an extensive list of dependencies for jupyter, which is why they suggest `pip` or `conda` on their own installation page: http://jupyter.readthedocs.org/en/latest/install.html.

**Comment by amrit-poudel** [comment]
@blink1073 
I believe I have all the dependencies installed (please see my list above) for ipython to work, which is why it works fine if i launch it from the terminal. However, it does not work, if I try to un ipython from spyder.

**Comment by blink1073** [comment]
I'm sorry, but there are already too many configurations (operating systems, python versions), to also debug non-standard installs.  I agree with @ccordoba12.

**Comment by amrit-poudel** [comment]
That's okay. I will try to figure it out on my own. Thanks.

**Comment by MarcusStein** [comment]
I have exactly the same problem (also a custom Python installation from source).

@amrit-poudel : Did you manage to resolve the issue?

**Comment by amrit-poudel** [comment]
@MarcusStein ,

There was not an easy solution to this problem. 

First, I built ipython and all the dependencies listed on their website from the source. On the terminal, this installation of ipython worked fine with no error. However it would not run within the spyder IDE.

I think the ipython website does not really list *all* dependencies for ipython. It lists the most critical ones that allows you to build ipython from source and run it from the terminal without any issue. However, this bare bone built of ipython is not sufficient to work seamlessly within spyder IDE.

Having no idea what other dependencies could be, I decided to pip install ipython (not  qtconsole or jupyter or sypder as instructed above by some, but *only* ipython). The pip installation of ipython downloaded/installed whole bunch of new dependencies that were not included in my previous built of ipython from source.

After pip installing ipython only, I was able to run ipython from within the spyder, which was also built from source.

Long story short, try pip install ipython in a separate folder and compare  it with your ipython built from the source.  Hope this helps.
